db_stress_tool/no_batched_ops_stress.cc

```diff
@@ -805,14 +805,25 @@ class NonBatchedOpsStressTest : public StressTest {
     int64_t key_base = rand_keys[0];
     int column_family = rand_column_families[0];
     std::vector<std::unique_ptr<MutexLock>> range_locks;
+    range_locks.reserve(FLAGS_ingest_external_file_width);
+    std::vector<int64_t> keys;
+    keys.reserve(FLAGS_ingest_external_file_width);
     std::vector<uint32_t> values;
+    values.reserve(FLAGS_ingest_external_file_width);
     SharedState* shared = thread->shared;
 
+    assert(FLAGS_nooverwritepercent < 100);
     // Grab locks, set pending state on expected values, and add keys
     for (int64_t key = key_base;
-         s.ok() && key < std::min(key_base + FLAGS_ingest_external_file_width,
-                                  shared->GetMaxKey());
+         s.ok() && key < shared->GetMaxKey() &&
+         static_cast<int32_t>(keys.size()) < FLAGS_ingest_external_file_width;
          ++key) {
+      if (!shared->AllowsOverwrite(key)) {
+        // We could alternatively grab `key`'s lock and then include it on the
+        // condition its current value is `DELETION_SENTINEL`.
+        continue;
+      }
+      keys.push_back(key);
       if (key == key_base) {
         range_locks.emplace_back(std::move(lock));
```

Comment:
If `key` is between `key_base` and the next key such that `next_key & ((1 << FLAGS_log2_keys_per_lock) - 1) == 0`, we might still need to do `range_locks.emplace_back(std::move(lock));` ?

Reply:
Good point, this would have been really hard to debug from a crash test failure - thanks! I think the problem also applies to the other `range_locks.emplace_back()` below -- we could `continue` past the first key in a lock's range and then simply not acquire the lock, while other keys in that lock's range could still be mutated.

Attempted to fix by moving the `continue` below all the lock acquisition.

Reply:
Ah yes, I did not realize it applies to the other lock statement too